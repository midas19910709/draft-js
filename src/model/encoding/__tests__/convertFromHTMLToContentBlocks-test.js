/**
 * Copyright (c) 2013-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @emails oncall+ui_infra
 * @format
 */

'use strict';

jest.disableAutomock();

jest.mock('generateRandomKey');

const DefaultDraftBlockRenderMap = require('DefaultDraftBlockRenderMap');

const convertFromHTMLToContentBlocks = require('convertFromHTMLToContentBlocks');
const getSafeBodyFromHTML = require('getSafeBodyFromHTML');

const DEFAULT_CONFIG = {
  DOMBuilder: getSafeBodyFromHTML,
  blockRenderMap: DefaultDraftBlockRenderMap,
  experimentalTreeDataSupport: false,
};

const IMAGE_DATA_URL =
  'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///' +
  'yH5BAEAAAAALAAAAAABAAEAAAIBRAA7';

const SUPPORTED_TAGS = [
  'blockquote',
  'div',
  'figure',
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'li',
  'p',
  'pre',
];

const normalizeBlock = block => {
  const {type, depth, text, characterList} = block;

  return {
    type,
    depth,
    text,
    characterList,
  };
};

const toggleExperimentalTreeDataSupport = enabled => {
  jest.doMock('DraftFeatureFlags', () => {
    return {
      draft_tree_data_support: enabled,
    };
  });
};

beforeEach(() => {
  jest.resetModules();
});

const convertFromHTML = (html_string, config) => {
  const options = {
    ...DEFAULT_CONFIG,
    ...config,
  };

  const {DOMBuilder, blockRenderMap, experimentalTreeDataSupport} = options;

  jest.resetModules();
  toggleExperimentalTreeDataSupport(experimentalTreeDataSupport);
  return convertFromHTMLToContentBlocks(
    html_string,
    DOMBuilder,
    blockRenderMap,
  );
};

const AreTreeBlockNodesEquivalent = (html_string, config = {}) => {
  const treeEnabled = convertFromHTML(html_string, {
    ...config,
    experimentalTreeDataSupport: true,
  }).contentBlocks.map(block => normalizeBlock(block.toJS()));

  const treeDisabled = convertFromHTML(html_string, {
    ...config,
    experimentalTreeDataSupport: false,
  }).contentBlocks.map(block => normalizeBlock(block.toJS()));

  return JSON.stringify(treeEnabled) === JSON.stringify(treeDisabled);
};

const assertConvertFromHTMLToContentBlocks = (html_string, config = {}) => {
  expect(
    convertFromHTML(html_string, config).contentBlocks.map(block =>
      block.toJS(),
    ),
  ).toMatchSnapshot();
};

const testConvertingAdjacentHtmlElementsToContentBlocks = (
  tag: string,
  experimentalTreeDataSupport?: boolean = false,
) => {
  test(`must not merge tags when converting adjacent <${tag} />`, () => {
    const html_string = `
      <${tag}>a</${tag}>
      <${tag}>b</${tag}>
    `;

    assertConvertFromHTMLToContentBlocks(
      html_string,
      experimentalTreeDataSupport,
    );
  });
};

const testConvertingHtmlElementsToContentBlocksAndRootContentBlockNodesMatch = (
  tag: string,
) => {
  test(`must convert root ContentBlockNodes to matching ContentBlock nodes for <${tag} />`, () => {
    expect(
      AreTreeBlockNodesEquivalent(`<${tag}>a</${tag}> `),
    ).toMatchSnapshot();
  });
};

SUPPORTED_TAGS.forEach(tag =>
  testConvertingAdjacentHtmlElementsToContentBlocks(tag),
);

test('img with http protocol should have camera emoji content', () => {
  const blocks = convertFromHTMLToContentBlocks(
    '<img src="http://www.facebook.com">',
  );
  expect(blocks.contentBlocks[0].text).toMatchSnapshot();
});

test('img with data protocol should be correctly parsed', () => {
  const blocks = convertFromHTMLToContentBlocks(
    `<img src="${IMAGE_DATA_URL}">`,
  );
  expect(blocks.contentBlocks[0].text).toMatchSnapshot();
});

test('converts nested html blocks when experimentalTreeDataSupport is enabled', () => {
  const html_string = `
    <blockquote>
      <h1>Hello World!</h1>
      <p>lorem ipsum</p>
    </blockquote>
  `;

  assertConvertFromHTMLToContentBlocks(html_string, {
    experimentalTreeDataSupport: true,
  });
});

test('converts deeply nested html blocks when experimentalTreeDataSupport is enabled', () => {
  const html_string = `
    <ol>
      <li>Some quote</li>
      <li>
        <blockquote>
          <h1>Hello World!</h1>
          <p>lorem ipsum</p>
        </blockquote>
      </li>
    </ol>
  `;

  assertConvertFromHTMLToContentBlocks(html_string, {
    experimentalTreeDataSupport: true,
  });
});

test('does not convert deeply nested html blocks when experimentalTreeDataSupport is disabled', () => {
  const html_string = `
    <ol>
      <li>Some quote</li>
      <li>
        <blockquote>
          <h1>Hello World!</h1>
          <p>lorem ipsum</p>
        </blockquote>
      </li>
    </ol>
  `;

  expect(AreTreeBlockNodesEquivalent(html_string)).toMatchSnapshot();
  assertConvertFromHTMLToContentBlocks(html_string, {
    experimentalTreeDataSupport: false,
  });
});

SUPPORTED_TAGS.forEach(tag =>
  testConvertingAdjacentHtmlElementsToContentBlocks(tag, true),
);

// assert that using tree blocks and root content block nodes are equivalent
SUPPORTED_TAGS.forEach(tag =>
  testConvertingHtmlElementsToContentBlocksAndRootContentBlockNodesMatch(tag),
);
