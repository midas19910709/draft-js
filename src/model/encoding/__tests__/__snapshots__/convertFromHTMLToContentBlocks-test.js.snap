// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converts deeply nested html blocks when experimentalTreeDataSupport is enabled 1`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "children": Array [],
    "data": Object {},
    "depth": 0,
    "key": "key5",
    "nextSibling": null,
    "parent": undefined,
    "prevSibling": null,
    "text": " ",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "children": Array [],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "nextSibling": "key1",
    "parent": null,
    "prevSibling": null,
    "text": "Some quote ",
    "type": "ordered-list-item",
  },
  Object {
    "characterList": Array [],
    "children": Array [
      "key2",
    ],
    "data": Object {},
    "depth": 0,
    "key": "key1",
    "nextSibling": null,
    "parent": null,
    "prevSibling": "key0",
    "text": "",
    "type": "ordered-list-item",
  },
  Object {
    "characterList": Array [],
    "children": Array [
      "key3",
      "key4",
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "nextSibling": null,
    "parent": "key1",
    "prevSibling": null,
    "text": "",
    "type": "blockquote",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "children": Array [],
    "data": Object {},
    "depth": 0,
    "key": "key3",
    "nextSibling": "key4",
    "parent": "key2",
    "prevSibling": null,
    "text": "Hello World! ",
    "type": "header-one",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "children": Array [],
    "data": Object {},
    "depth": 0,
    "key": "key4",
    "nextSibling": null,
    "parent": "key2",
    "prevSibling": "key3",
    "text": "lorem ipsum 
  ",
    "type": "unstyled",
  },
]
`;

exports[`converts nested html blocks when experimentalTreeDataSupport is enabled 1`] = `
Array [
  Object {
    "characterList": Array [],
    "children": Array [
      "key1",
      "key2",
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "nextSibling": null,
    "parent": null,
    "prevSibling": null,
    "text": "",
    "type": "blockquote",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "children": Array [],
    "data": Object {},
    "depth": 0,
    "key": "key1",
    "nextSibling": "key2",
    "parent": "key0",
    "prevSibling": null,
    "text": "Hello World!
 ",
    "type": "header-one",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "children": Array [],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "nextSibling": null,
    "parent": "key0",
    "prevSibling": "key1",
    "text": "lorem ipsum
 ",
    "type": "unstyled",
  },
]
`;

exports[`does not convert deeply nested html blocks when experimentalTreeDataSupport is disabled 1`] = `false`;

exports[`does not convert deeply nested html blocks when experimentalTreeDataSupport is disabled 2`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key4",
    "text": " ",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "Some quote",
    "type": "ordered-list-item",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "Hello World!
 lorem ipsum
 
 ",
    "type": "ordered-list-item",
  },
]
`;

exports[`img with data protocol should be correctly parsed 1`] = `"📷"`;

exports[`img with http protocol should have camera emoji content 1`] = `"📷"`;

exports[`must convert root ContentBlockNodes to matching ContentBlock nodes for <blockquote /> 1`] = `true`;

exports[`must convert root ContentBlockNodes to matching ContentBlock nodes for <div /> 1`] = `true`;

exports[`must convert root ContentBlockNodes to matching ContentBlock nodes for <figure /> 1`] = `true`;

exports[`must convert root ContentBlockNodes to matching ContentBlock nodes for <h1 /> 1`] = `true`;

exports[`must convert root ContentBlockNodes to matching ContentBlock nodes for <h2 /> 1`] = `true`;

exports[`must convert root ContentBlockNodes to matching ContentBlock nodes for <h3 /> 1`] = `true`;

exports[`must convert root ContentBlockNodes to matching ContentBlock nodes for <h4 /> 1`] = `true`;

exports[`must convert root ContentBlockNodes to matching ContentBlock nodes for <h5 /> 1`] = `true`;

exports[`must convert root ContentBlockNodes to matching ContentBlock nodes for <h6 /> 1`] = `true`;

exports[`must convert root ContentBlockNodes to matching ContentBlock nodes for <li /> 1`] = `true`;

exports[`must convert root ContentBlockNodes to matching ContentBlock nodes for <p /> 1`] = `true`;

exports[`must convert root ContentBlockNodes to matching ContentBlock nodes for <pre /> 1`] = `true`;

exports[`must not merge tags when converting adjacent <blockquote /> 1`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "blockquote",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "blockquote",
  },
]
`;

exports[`must not merge tags when converting adjacent <blockquote /> 2`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "blockquote",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "blockquote",
  },
]
`;

exports[`must not merge tags when converting adjacent <div /> 1`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "unstyled",
  },
]
`;

exports[`must not merge tags when converting adjacent <div /> 2`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "unstyled",
  },
]
`;

exports[`must not merge tags when converting adjacent <figure /> 1`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "atomic",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "atomic",
  },
]
`;

exports[`must not merge tags when converting adjacent <figure /> 2`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "atomic",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "atomic",
  },
]
`;

exports[`must not merge tags when converting adjacent <h1 /> 1`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "header-one",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "header-one",
  },
]
`;

exports[`must not merge tags when converting adjacent <h1 /> 2`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "header-one",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "header-one",
  },
]
`;

exports[`must not merge tags when converting adjacent <h2 /> 1`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "header-two",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "header-two",
  },
]
`;

exports[`must not merge tags when converting adjacent <h2 /> 2`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "header-two",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "header-two",
  },
]
`;

exports[`must not merge tags when converting adjacent <h3 /> 1`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "header-three",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "header-three",
  },
]
`;

exports[`must not merge tags when converting adjacent <h3 /> 2`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "header-three",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "header-three",
  },
]
`;

exports[`must not merge tags when converting adjacent <h4 /> 1`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "header-four",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "header-four",
  },
]
`;

exports[`must not merge tags when converting adjacent <h4 /> 2`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "header-four",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "header-four",
  },
]
`;

exports[`must not merge tags when converting adjacent <h5 /> 1`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "header-five",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "header-five",
  },
]
`;

exports[`must not merge tags when converting adjacent <h5 /> 2`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "header-five",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "header-five",
  },
]
`;

exports[`must not merge tags when converting adjacent <h6 /> 1`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "header-six",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "header-six",
  },
]
`;

exports[`must not merge tags when converting adjacent <h6 /> 2`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "header-six",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "header-six",
  },
]
`;

exports[`must not merge tags when converting adjacent <li /> 1`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "unordered-list-item",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "unordered-list-item",
  },
]
`;

exports[`must not merge tags when converting adjacent <li /> 2`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "unordered-list-item",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "unordered-list-item",
  },
]
`;

exports[`must not merge tags when converting adjacent <p /> 1`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "unstyled",
  },
]
`;

exports[`must not merge tags when converting adjacent <p /> 2`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "unstyled",
  },
]
`;

exports[`must not merge tags when converting adjacent <pre /> 1`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "code-block",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "code-block",
  },
]
`;

exports[`must not merge tags when converting adjacent <pre /> 2`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key0",
    "text": "a",
    "type": "code-block",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "key2",
    "text": "b",
    "type": "code-block",
  },
]
`;
